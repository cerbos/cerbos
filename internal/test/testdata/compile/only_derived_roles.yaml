---
mainDef: "derived_roles/my_derived_roles.yaml"
inputDefs:
  "derived_roles/my_derived_roles.yaml":
    apiVersion: "api.cerbos.dev/v1"
    derived_roles:
      name: my_derived_roles
      definitions:
        - name: tester
          parentRoles: ["dev", "qa"]

        - name: employee_that_owns_the_record
          parentRoles: ["employee"]
          computation:
            script: |-
              input.resource.attr.owner == input.principal.id

        - name: any_employee
          parentRoles: ["employee"]

        - name: direct_manager
          parentRoles: ["manager"]
          computation:
            match:
              expr:
                - "request.resource.attr.geography == request.principal.attr.geography"
                - "request.resource.attr.geography == request.principal.attr.managed_geographies"
