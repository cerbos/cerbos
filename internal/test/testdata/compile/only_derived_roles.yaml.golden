{
  "fqn":  "cerbos.derived_roles.my_derived_roles",
  "derivedRoles":  {
    "meta":  {
      "fqn":  "cerbos.derived_roles.my_derived_roles"
    },
    "derivedRoles":  {
      "any_employee":  {
        "name":  "any_employee",
        "parentRoles":  {
          "employee":  {}
        }
      },
      "direct_manager":  {
        "name":  "direct_manager",
        "parentRoles":  {
          "manager":  {}
        },
        "condition":  {
          "all":  {
            "expr":  [
              {
                "expr":  {
                  "original":  "request.resource.attr.geography == request.principal.attr.geography",
                  "checked":  {
                    "referenceMap":  {
                      "1":  {
                        "name":  "request"
                      },
                      "5":  {
                        "overloadId":  [
                          "equals"
                        ]
                      },
                      "6":  {
                        "name":  "request"
                      }
                    },
                    "typeMap":  {
                      "1":  {
                        "messageType":  "cerbos.engine.v1.CheckInput"
                      },
                      "2":  {
                        "messageType":  "cerbos.engine.v1.Resource"
                      },
                      "3":  {
                        "mapType":  {
                          "keyType":  {
                            "primitive":  "STRING"
                          },
                          "valueType":  {
                            "dyn":  {}
                          }
                        }
                      },
                      "4":  {
                        "dyn":  {}
                      },
                      "5":  {
                        "primitive":  "BOOL"
                      },
                      "6":  {
                        "messageType":  "cerbos.engine.v1.CheckInput"
                      },
                      "7":  {
                        "messageType":  "cerbos.engine.v1.Principal"
                      },
                      "8":  {
                        "mapType":  {
                          "keyType":  {
                            "primitive":  "STRING"
                          },
                          "valueType":  {
                            "dyn":  {}
                          }
                        }
                      },
                      "9":  {
                        "dyn":  {}
                      }
                    },
                    "sourceInfo":  {
                      "location":  "<input>",
                      "lineOffsets":  [
                        68
                      ],
                      "positions":  {
                        "1":  0,
                        "2":  7,
                        "3":  16,
                        "4":  21,
                        "5":  32,
                        "6":  35,
                        "7":  42,
                        "8":  52,
                        "9":  57
                      }
                    },
                    "expr":  {
                      "id":  "5",
                      "callExpr":  {
                        "function":  "_==_",
                        "args":  [
                          {
                            "id":  "4",
                            "selectExpr":  {
                              "operand":  {
                                "id":  "3",
                                "selectExpr":  {
                                  "operand":  {
                                    "id":  "2",
                                    "selectExpr":  {
                                      "operand":  {
                                        "id":  "1",
                                        "identExpr":  {
                                          "name":  "request"
                                        }
                                      },
                                      "field":  "resource"
                                    }
                                  },
                                  "field":  "attr"
                                }
                              },
                              "field":  "geography"
                            }
                          },
                          {
                            "id":  "9",
                            "selectExpr":  {
                              "operand":  {
                                "id":  "8",
                                "selectExpr":  {
                                  "operand":  {
                                    "id":  "7",
                                    "selectExpr":  {
                                      "operand":  {
                                        "id":  "6",
                                        "identExpr":  {
                                          "name":  "request"
                                        }
                                      },
                                      "field":  "principal"
                                    }
                                  },
                                  "field":  "attr"
                                }
                              },
                              "field":  "geography"
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              },
              {
                "expr":  {
                  "original":  "request.resource.attr.geography == request.principal.attr.managed_geographies",
                  "checked":  {
                    "referenceMap":  {
                      "1":  {
                        "name":  "request"
                      },
                      "5":  {
                        "overloadId":  [
                          "equals"
                        ]
                      },
                      "6":  {
                        "name":  "request"
                      }
                    },
                    "typeMap":  {
                      "1":  {
                        "messageType":  "cerbos.engine.v1.CheckInput"
                      },
                      "2":  {
                        "messageType":  "cerbos.engine.v1.Resource"
                      },
                      "3":  {
                        "mapType":  {
                          "keyType":  {
                            "primitive":  "STRING"
                          },
                          "valueType":  {
                            "dyn":  {}
                          }
                        }
                      },
                      "4":  {
                        "dyn":  {}
                      },
                      "5":  {
                        "primitive":  "BOOL"
                      },
                      "6":  {
                        "messageType":  "cerbos.engine.v1.CheckInput"
                      },
                      "7":  {
                        "messageType":  "cerbos.engine.v1.Principal"
                      },
                      "8":  {
                        "mapType":  {
                          "keyType":  {
                            "primitive":  "STRING"
                          },
                          "valueType":  {
                            "dyn":  {}
                          }
                        }
                      },
                      "9":  {
                        "dyn":  {}
                      }
                    },
                    "sourceInfo":  {
                      "location":  "<input>",
                      "lineOffsets":  [
                        78
                      ],
                      "positions":  {
                        "1":  0,
                        "2":  7,
                        "3":  16,
                        "4":  21,
                        "5":  32,
                        "6":  35,
                        "7":  42,
                        "8":  52,
                        "9":  57
                      }
                    },
                    "expr":  {
                      "id":  "5",
                      "callExpr":  {
                        "function":  "_==_",
                        "args":  [
                          {
                            "id":  "4",
                            "selectExpr":  {
                              "operand":  {
                                "id":  "3",
                                "selectExpr":  {
                                  "operand":  {
                                    "id":  "2",
                                    "selectExpr":  {
                                      "operand":  {
                                        "id":  "1",
                                        "identExpr":  {
                                          "name":  "request"
                                        }
                                      },
                                      "field":  "resource"
                                    }
                                  },
                                  "field":  "attr"
                                }
                              },
                              "field":  "geography"
                            }
                          },
                          {
                            "id":  "9",
                            "selectExpr":  {
                              "operand":  {
                                "id":  "8",
                                "selectExpr":  {
                                  "operand":  {
                                    "id":  "7",
                                    "selectExpr":  {
                                      "operand":  {
                                        "id":  "6",
                                        "identExpr":  {
                                          "name":  "request"
                                        }
                                      },
                                      "field":  "principal"
                                    }
                                  },
                                  "field":  "attr"
                                }
                              },
                              "field":  "managed_geographies"
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              }
            ]
          }
        }
      },
      "employee_that_owns_the_record":  {
        "name":  "employee_that_owns_the_record",
        "parentRoles":  {
          "employee":  {}
        },
        "condition":  {
          "expr":  {
            "original":  "request.resource.attr.owner == request.principal.id",
            "checked":  {
              "referenceMap":  {
                "1":  {
                  "name":  "request"
                },
                "5":  {
                  "overloadId":  [
                    "equals"
                  ]
                },
                "6":  {
                  "name":  "request"
                }
              },
              "typeMap":  {
                "1":  {
                  "messageType":  "cerbos.engine.v1.CheckInput"
                },
                "2":  {
                  "messageType":  "cerbos.engine.v1.Resource"
                },
                "3":  {
                  "mapType":  {
                    "keyType":  {
                      "primitive":  "STRING"
                    },
                    "valueType":  {
                      "dyn":  {}
                    }
                  }
                },
                "4":  {
                  "dyn":  {}
                },
                "5":  {
                  "primitive":  "BOOL"
                },
                "6":  {
                  "messageType":  "cerbos.engine.v1.CheckInput"
                },
                "7":  {
                  "messageType":  "cerbos.engine.v1.Principal"
                },
                "8":  {
                  "primitive":  "STRING"
                }
              },
              "sourceInfo":  {
                "location":  "<input>",
                "lineOffsets":  [
                  52
                ],
                "positions":  {
                  "1":  0,
                  "2":  7,
                  "3":  16,
                  "4":  21,
                  "5":  28,
                  "6":  31,
                  "7":  38,
                  "8":  48
                }
              },
              "expr":  {
                "id":  "5",
                "callExpr":  {
                  "function":  "_==_",
                  "args":  [
                    {
                      "id":  "4",
                      "selectExpr":  {
                        "operand":  {
                          "id":  "3",
                          "selectExpr":  {
                            "operand":  {
                              "id":  "2",
                              "selectExpr":  {
                                "operand":  {
                                  "id":  "1",
                                  "identExpr":  {
                                    "name":  "request"
                                  }
                                },
                                "field":  "resource"
                              }
                            },
                            "field":  "attr"
                          }
                        },
                        "field":  "owner"
                      }
                    },
                    {
                      "id":  "8",
                      "selectExpr":  {
                        "operand":  {
                          "id":  "7",
                          "selectExpr":  {
                            "operand":  {
                              "id":  "6",
                              "identExpr":  {
                                "name":  "request"
                              }
                            },
                            "field":  "principal"
                          }
                        },
                        "field":  "id"
                      }
                    }
                  ]
                }
              }
            }
          }
        }
      },
      "tester":  {
        "name":  "tester",
        "parentRoles":  {
          "dev":  {},
          "qa":  {}
        }
      }
    }
  }
}